livre/lib/cache/HistogramObject.cpp

```diff
@@ -0,0 +1,240 @@
+                                             compCount * j * blockSize.z() +
+                                             compCount * k + c;
+                        const uint8_t data = rawData[ index ];
+                        bins[ data ] += scaleFactor;
+                    }
+    }
+
+    template< class SRC_TYPE >
+    void binData( const SRC_TYPE* rawData,
+                  Histogram& histogram,
+                  const Vector3ui& blockSize,
+                  const Vector3ui& padding,
+                  const size_t compCount,
+                  const uint64_t scaleFactor ) const
+    {
+        uint64_t* bins = histogram.getBins();
+        const float minVal = std::numeric_limits< SRC_TYPE >::min();
+        const float maxVal = std::numeric_limits< SRC_TYPE >::max();
+        const float range = maxVal - minVal;
+
+        for( size_t i = padding.x(); i < blockSize.x() - padding.x(); ++i )
+            for( size_t j = padding.y(); j < blockSize.y() - padding.y(); ++j )
+                for( size_t k = padding.z(); k < blockSize.z() -padding.z(); ++k )
+                    for( size_t c = 0; c < compCount; ++c )
+                    {
+                        const size_t index = compCount * i * blockSize.y() * blockSize.z() +
+                                             compCount * j * blockSize.z() +
+                                             compCount * k + c;
+                        const float data = rawData[ index ];
+                        const size_t binIndex =
+                                std::is_unsigned< SRC_TYPE >::value ?
+                                ( data - minVal ) / range * Histogram::binCount :
+                                (( data / range ) + 0.5f ) * Histogram::binCount;
+                        bins[ binIndex ] += scaleFactor;
+                    }
+    }
+
+    bool load()
+    {
+        const VolumeInformation& volumeInfo = _dataCache.getDataSource().getVolumeInfo();
+        const size_t compCount = volumeInfo.compCount;
+
+        if( compCount > 1 )
+            LBTHROW( std::runtime_error( "Multiple channels are not supported "));
+
+        ConstTextureDataObjectPtr textureData =
+                std::static_pointer_cast< const TextureDataObject >(
+                    _dataCache.get( _histogramObject.getId( )));
+
+        if( !textureData->isLoaded( ))
+            return false;
+
+        const void* rawData = textureData->getDataPtr();
+        const LODNode& lodNode = _dataCache.getDataSource().getNode(
+                    NodeId( _histogramObject.getId( )));
+        const Vector3ui& voxelBox = lodNode.getVoxelBox().getSize();
+        const Vector3ui& padding = volumeInfo.overlap;
+
+        const uint64_t scaleFactor1d =
+                1 << ( volumeInfo.rootNode.getDepth() - lodNode.getRefLevel() - 1 );
+
+        const uint64_t scaleFactor = scaleFactor1d * scaleFactor1d * scaleFactor1d;
```

Comment:
scaleFactor1d is a power of two, you can also do scaleFactor1d << 3 or even multiply  volumeInfo.rootNode.getDepth() - lodNode.getRefLevel() - 1 by 3 before the left shift.